datafusion/core/src/physical_plan/memory.rs

```diff
@@ -327,24 +326,25 @@ impl ExecutionPlan for MemoryWriteExec {
     ) -> Result<SendableRecordBatchStream> {
         let batch_count = self.batches.len();
         let data = self.input.execute(partition, context)?;
-        if batch_count >= self.input.output_partitioning().partition_count() {
-            // If the number of input partitions matches the number of MemTable partitions,
-            // use a lightweight implementation that doesn't utilize as many locks.
-            let table_partition = self.batches[partition].clone();
-            Ok(Box::pin(MemorySinkOneToOneStream::try_new(
-                table_partition,
-                data,
-                self.schema.clone(),
-            )?))
-        } else {
-            // Otherwise, use the locked implementation.
-            let table_partition = self.batches[partition % batch_count].clone();
-            Ok(Box::pin(MemorySinkStream::try_new(
-                table_partition,
-                data,
-                self.schema.clone(),
-            )?))
-        }
+        let schema = self.schema.clone();
+        let state = (data, self.batches[partition % batch_count].clone());
+
+        let stream = futures::stream::unfold(state, |mut state| async move {
+            // hold lock during the entire write
+            let mut locked = state.1.write().await;
+            loop {
+                let batch = match state.0.next().await {
+                    Some(Ok(batch)) => batch,
+                    Some(Err(e)) => {
+                        drop(locked);
+                        return Some((Err(e), state));
+                    }
+                    None => return None,
+                };
+                locked.push(batch)
+            }
+        });
+        Ok(Box::pin(RecordBatchStreamAdapter::new(schema, stream)))
```

Comment:
```suggestion
       let adapter = if batch_count >= self.input.output_partitioning().partition_count()
        {
            // If the number of input partitions matches the number of MemTable partitions,
            // use a lightweight implementation that doesn't utilize as many locks.
            let stream = futures::stream::unfold(state, |mut state| async move {
                // hold lock during the entire write
                let mut locked = state.1.write().await;
                loop {
                    let batch = match state.0.next().await {
                        Some(Ok(batch)) => batch,
                        Some(Err(e)) => {
                            drop(locked);
                            return Some((Err(e), state));
                        }
                        None => return None,
                    };
                    locked.push(batch)
                }
            });
            Box::pin(RecordBatchStreamAdapter::new(schema, stream))
                as SendableRecordBatchStream
        } else {
            let stream = futures::stream::unfold(state, |mut state| async move {
                loop {
                    let batch = match state.0.next().await {
                        Some(Ok(batch)) => batch,
                        Some(Err(e)) => return Some((Err(e), state)),
                        None => return None,
                    };
                    state.1.write().await.push(batch)
                }
            });
            Box::pin(RecordBatchStreamAdapter::new(schema, stream))
                as SendableRecordBatchStream
        };
        Ok(adapter)
```

Reply:
The situation where `batch_count` is less than the number of partitions from `self.input.output_partitioning().partition_count()` could potentially lead to unnecessary lock waiting when multiple input partitions are aimed at a single table partition.

Reply:
After thinking about this more, I don't understand why we are locking at all in the loop -- if we are really worried about the locking performance we can just buffer the batches and copy them at the end of the stream. This should also keep the code simpler.  I'll make a follow on PR with a proposal